src/nodes.coffee

```diff
@@ -1770,10 +1828,60 @@ exports.Assign = class Assign extends Base
     answer = compiledName.concat @makeCode(" #{ @context or '=' } "), val
     # Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assignment_without_declaration,
     # if we’re destructuring without declaring, the destructuring assignment must be wrapped in parentheses.
-    if o.level > LEVEL_LIST or (isValue and @variable.base instanceof Obj and not @param)
-      @wrapInParentheses answer
+    # Remove `answers.unshift()` parts When ES proposal for rest elements in object destructuring hits Stage 4.
+    if o.level > LEVEL_LIST or (isValue and @variable.base instanceof Obj and (not @param or @paramWithSplat))
+      answers.unshift @wrapInParentheses answer
+      # @wrapInParentheses answer
     else
-      answer
+      answers.unshift answer
+      # answer
+    @joinFragmentArrays [answersOnTop..., answers...], ', '
+
+  # Check object destructuring variable for rest elements;
+  # can be removed once ES proposal hits Stage 4.
+  compileObjectDestruct: (o) ->
+    getPropValue = (obj) ->
+      fragmentsToText (if obj instanceof Assign then obj.variable.unwrapAll() else obj.unwrap()).compileToFragments(o)
+    # Recursive function for searching and storing rest elements in objects.
+    # Parameter props[] is used to store nested object properties,
+    # e.g. `{a: {b, c: {d, r1...}, r2...}, r3...} = obj`.
+    traverseRest = (objects, props = []) ->
+      results = []
+      restElement = no
+      for obj, key in objects
+        if obj instanceof Assign and obj.context == "object" and obj.value.base instanceof Obj
+          props.push getPropValue obj
+          results = traverseRest obj.value.base.objects, props
+          props.pop()
+        if obj instanceof Splat
+          obj.error "multiple rest elements are disallowed in object destructuring" if restElement
+          restElement = {key, name: obj.unwrap(), props: [props...]}
+      if restElement
+        # Remove rest element from the object.
+        objects.splice restElement.key, 1
+        # Prepare array of property keys to be excluded from the object
+        # TODO: refactor
+        excludeProps = (getPropValue(obj) for obj in objects)
```

Comment:
Agree. The current solution has too much of a "here be dragons" feeling. Hence the "refactor" reminder.